exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -274,11 +274,8 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
 
 	var providerIDList []string
 	// Get Spec.ProviderIDList from the infrastructure provider.
-	if err := util.UnstructuredUnmarshalField(infraConfig, &providerIDList, "spec", "providerIDList"); err != nil {
+	if err := util.UnstructuredUnmarshalField(infraConfig, &providerIDList, "spec", "providerIDList"); err != nil && !errors.Is(err, util.ErrUnstructuredFieldNotFound) {
```

Comment:
Is there a case where we could get a `util.ErrUnstructuredFieldNotFound` from an actual error, i.e. we expect a `providerIDList`, but it gets suppressed here?

Reply:
Looking at the [AWS implementation](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/exp/controllers/awsmachinepool_controller.go#L273) and the [Azure implementation](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/d0b9621c43faca25266182fcfdcf585079688f06/azure/scope/machinepool.go#L222), the `providerIDList` should always be set if there are any instances in the respective pool implementation. 
`providerIDList` is in both cases marked with the JSON tag `,omitempty` so an empty list will result in the field being not set. 

The `util.UnstructuredUnmarshalField()` implementation returns `util.ErrUnstructuredFieldNotFound` only in the following cases:
1. infrastructure object contains no `spec` or is not a map
2. infrastructure object contains no `spec.providerIDList` or `spec.providerIDList` is nil

Considering this, I'd say this check should be ok. #1 might be an issue but then the infrastructure provider implementation has an issue. 

Bear in mind: I'm new to this whole codebase - correct me if I'm wrong please :)